superset/views/core.py

```diff
@@ -1788,25 +1788,40 @@ def dashboard(self, dashboard_id):
         dash = qry.one_or_none()
         if not dash:
             abort(404)
-        datasources = set()
+        datasources = {}
```

Comment:
Could we use a `defaultdict(lambda: defaultdict(list))` ? This means the logic in lines 1795 - 1801 could be simplified.

Reply:
i don't think this actually helps anything because we need to keep multiple fields in the dict of dicts, the datasource and a list of slices

Reply:
I don't believe the datasource can be the key as it's not hashable. I think your first solution will work though

Reply:
You can probably create a hash function in the datasource class?

```
def ___hash__(self):
    return self.uid
```

Reply:
i've added a `__hash__` and an `__eq__` function as according to https://docs.python.org/3/glossary.html#term-hashable